### ISSUE TITLE
Request for Assistance: Missing kexec-tools in Jammy/Focal Images for Integration Test Suite

### ISSUE THREAD

**Issue by smit-gardhariya** [cloud-hypervisor/cloud-hypervisor#6726]
I am currently working on a simple kexec test, which I intend to incorporate into our integration test suite. However, I have encountered an issue where the `kexec-tools` package is not installed by default on the Jammy and Focal base images. Since this package is essential for executing the test, I would like assistance in either pre-installing `kexec-tools` in the images or determining the best approach to ensure the package is available during the test run.

Also, if you can help us with instructions on how to update images being used by test-framework or on how we can use our own image, that would really be helpful. I can see it is being downloaded from predefined location which might be accessible for common user like us.

**Comment by smit-gardhariya** [comment]
@rbradford @liuw @jinankjain Could you please help me with above query. Thnx :)

**Comment by likebreath** [comment]
Please take a look at our custom-image documentation [1] about how you can extend the base image we are using for our CI. 
The commit a86edd3032664a2e2d9c6436696ea6d4d1e55aba is a good reference. Of course, we can help with uploading the updated image to our Azure storage account.


[1] https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/custom-image.md

**Comment by smit-gardhariya** [comment]
@likebreath Thank you for sharing details. This is really helpful.

I can create image but how do i upload. I might not have access to storage account. Also document you mentioned above, has some reference of web interface for upload but did not say much. Is the same azure portal or some other web interface to upload file.

**Comment by rbradford** [comment]
Could you elaborate - are you planning on adding an integration test to the Cloud Hypervisor test suite - if so do you have a branch we can look at? We will happily add new packages to our CI image if new tests require them but if this is just for your own purpose you should build your own image to use.

**Comment by smit-gardhariya** [comment]
@rbradford Yes i am planning to add testcase with integration test. As of now it is simple kexec test which will try to load new kernel with some additional param set in kernel command line and post-kexec it will read `/proc/cmdline` to verify if new param set correctly. 

Basic requirement for this is to 

- Install `kexec-tools` on the guest image.
- Enable kernel config for kexec.

Test is using bzImage to create VM and then scp the same bzImage onto the guest. Later, testcase performed kexec as stated above and verify the kernel cmdline.

I have code changes with my fork but here is the test to look. For testing purpose, i did create my custom image and disabled the sha1sum check in `run_integration_tests_x86.sh`. 

```
    #[test]
    #[cfg(target_arch = "x86_64")]
    fn test_simple_kexec() {
        let jammy = UbuntuDiskConfig::new(JAMMY_IMAGE_NAME.to_string());
        let guest = Guest::new(Box::new(jammy));
        let mut kernel_path = direct_kernel_boot_path();
        kernel_path.pop();
        kernel_path.push("bzImage");
        let mut child = GuestCommand::new(&guest)
            .args(["--cpus", "boot=1"])
            .args(["--memory", "size=512M"])
            .args(["--kernel", kernel_path.to_str().unwrap()])
            .args(["--cmdline", DIRECT_KERNEL_BOOT_CMDLINE])
            .args(["--console", "tty"])
            .default_disks()
            .default_net()
            .capture_output()
            .spawn()
            .unwrap();
        let r = std::panic::catch_unwind(|| {
            guest.wait_vm_boot(None).unwrap();
            let bzimage_remote_path = "/home/cloud/bzImage";
            let _ = scp_to_guest(
                Path::new(kernel_path.to_str().unwrap()),
                Path::new(bzimage_remote_path),
                &guest.network.guest_ip,
                DEFAULT_SSH_RETRIES,
                DEFAULT_SSH_TIMEOUT,
            );
            let append_arg = "test_simple_kexec=1";
            let guest_command = format!(
                "sudo kexec -s {} --reuse-cmdline --append={}",
                &bzimage_remote_path, append_arg,
            );
            let _ = guest.ssh_command(guest_command.as_str()).unwrap();
            guest.wait_vm_boot(None).unwrap();
            let stdout_cmdline = guest.ssh_command("sudo cat /proc/cmdline").unwrap();
            let op = stdout_cmdline.trim();
            if !op.contains(append_arg) {
                panic!("New kernel cmdline is wrong");
            }
        });
        let _ = child.kill();
        let output = child.wait_with_output().unwrap();
        handle_child_output(r, &output);
    }
```

**Comment by smit-gardhariya** [comment]
@rbradford Here is the branch with my fork: https://github.com/smit-gardhariya/cloud-hypervisor/tree/smit-gardhariya/kexec_test

**Comment by rbradford** [comment]
Could you elaborate on your rationale as to why you think we need to test kexec functionality in Cloud Hypervisor  (full disclosure I know of kexec but never used it.) Is there something VMM specific there?

**Comment by liuw** [comment]
> Could you elaborate on your rationale as to why you think we need to test kexec functionality in Cloud Hypervisor (full disclosure I know of kexec but never used it.) Is there something VMM specific there?

Smit is a member of my team. We use kexec to service the VM (either for replacing the kernel, or for gathering crashdump when the VM crashes). The guest transitions through several different states (including CPU states and device states). If this is not tested, then anything changed in those areas in CH can potentially break kexec.

**Comment by smit-gardhariya** [comment]
@rbradford Test i mentioned is just for loading new kernel with kexec. I want to add the crash dump test as well where we may require other packages too. I will try some manual run and update this thread on needed packages.

Let me know your opinion on this too.

**Comment by smit-gardhariya** [comment]
@rbradford 

I saw Jammy being updated with packages. Could you please confirm for crash and kdump utility related packages too. So we can grab the needed packages in one go.

I think there are these packages kexec-tools and kdump-tools for canonical need to test the kexec/kdump tests. 

CC @jinankjain Please suggest any package need to be added.

**Comment by smit-gardhariya** [comment]
@rbradford, I tested the image `jammy-server-cloudimg-amd64-custom-20241017-0.qcow2` and it works fine for simple kexec test i mentioned above.

Could you please help me adding the following packages too. This would help adding test for crash dump as well.

crash
kdump-tools
linux-crashdump

CC @jinankjain

**Comment by TimePrinciple** [comment]
I can help here if others are busy, please assign me if that is the case :)

**Comment by smit-gardhariya** [comment]
@TimePrinciple Sure, if you can add packages i mentioned above in image `jammy-server-cloudimg-amd64-custom-20241017-0.qcow2` that would really help.

Rob created this image by adding kexec-tools (in case if you are creating new version altogether)

**Comment by rbradford** [comment]
@smit-gardhariya It's ambiguous to me if you need those packages for the integration test or not?

The custom image should only have the packages required by the integration tests - you can however easily create your own image now using this script - https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/scripts/build-custom-image.sh

**Comment by smit-gardhariya** [comment]
Sure Rob, let me check with this script.

These packages are required for new test i am working on to verify if crash dump is generating and crash utility can load it properly.

**Comment by liuw** [comment]
@smit-gardhariya once all the utilities are installed, it also makes sense to add a test case for dumping the guest with the VMM side crash dump.

**Comment by smit-gardhariya** [comment]
@rbradford It seems like CH v42 is now downloading the guest kernel instead of building it. Should we add the mechanism to tell CH testframe work whether it need to build it or use the pre-compiled binaries

Something like, 

To build guest kernel
`./scripts/dev_cli.sh tests --hypervisor mshv **--build_guest_kernel** --integration -- -- --exact common_parallel::test_kernel_crash_dump`

To Use precompiled binaries
`./scripts/dev_cli.sh tests --hypervisor mshv --integration -- -- --exact common_parallel::test_kernel_crash_dump`

CC @jinankjain

**Comment by TimePrinciple** [comment]
> @rbradford It seems like CH v42 is now downloading the guest kernel instead of building it. Should we add the mechanism to tell CH test framework whether it need to build it or use the pre-compiled binaries
> 
> I need to enable some kernel config for kernel crash to work and generate dump file.
> 
> Something like,
> 
> To build guest kernel `./scripts/dev_cli.sh tests --hypervisor mshv --build_guest_kernel --integration -- -- --exact common_parallel::test_kernel_crash_dump`
> 
> To Use precompiled binaries `./scripts/dev_cli.sh tests --hypervisor mshv --integration -- -- --exact common_parallel::test_kernel_crash_dump`
> 
> CC @jinankjain

If this is scheduled, I can pick it up - When Rob says so 👀

**Comment by rbradford** [comment]
Is this issue resolved now?

**Comment by smit-gardhariya** [comment]
Yes i am able to get the packages i need now. Closing the issue.
